src/main/java/com/emc/ecs/zimbra/integration/EcsStoreManager.java

```diff
@@ -184,15 +200,38 @@ public InputStream readStreamFromStore(String locator, Mailbox mbox) throws IOEx
     public boolean deleteFromStore(String locator, Mailbox mbox) throws IOException {
         EcsLogger.debug(String.format("deleteFromStore() - start: locator - %s, accountId - %s", locator, mbox.getId()));
 
-        EcsLocator el = EcsLocatorUtil.fromStringLocator(locator);
+        final EcsLocator el = EcsLocatorUtil.fromStringLocator(locator);
 
         EcsLogger.debug(String.format("deleteFromStore() - deleting: bucket - %s, key - %s", el.getBucketName(), el.getKey()));
-        try {
-            client.deleteObject(el.getBucketName(), el.getKey());
-        } catch (Exception e) {
-            EcsLogger.error(String.format("Failed to delete from - %s", locator), e);
-            return false;
-        }
+
+        boolean waiting = true;

```

Comment:
this whole block is going to waste a lot of CPU resources by checking the futures list every millisecond... unfortunately there is no easy way to approach this.  You could use a boundless job queue, but in the case of many deletes (millions), it would eat up tons of memory.  My best recommendation is to use the EnhancedThreadPoolExecutor class that I wrote for ecs-sync.  Use the `.blockingSubmit()` method to submit tasks and it will simply wait until the queue has space.
https://github.com/EMCECS/ecs-sync/blob/master/src/main/java/com/emc/ecs/sync/util/EnhancedThreadPoolExecutor.java
One caveat: you will have to delete the `newTaskFor()` methods to remove the dependency on EnhancedFutureTask.